Updating Token Hook Guide to single-page (#2847)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/token-inline-hook/main/index.md

---
title: Token Inline Hook
excerpt: Learn how to easily implement a Token Inline hook
layout: Guides
---

<StackSelector />

> This guide provides a working example of an Okta Token Inline Hook. It uses the website [Glitch.com](https://glitch.com) to act as an external service and receive and respond to Token Inline Hook calls.

Reviewer: ```suggestion
This guide provides a working example of an Okta Token Inline Hook. It uses the website [Glitch.com](https://glitch.com) to act as an external service to receive and respond to Token Inline Hook calls.
```

---

**Learning outcomes**

* Understand the Okta Token Inline Hook calls and responses.

> * Implement a simple working example of a Token Inline Hook with a Glitch.com project, which acts as an external service.

Reviewer: ```suggestion
* Implement a simple working example of a Token Inline Hook with a Glitch.com external service.
```

> * Test the Token Inline hook.

Reviewer: ```suggestion
* Test the Token Inline Hook.
```
"Hook" was previously capitalized, so I'm following the convention.

Author: Good catch. Thanks.

**What you need**

* A [Glitch.com](https://glitch.com) project or account to act as an external service.
* An Okta developer org. [Create an org for free](https://developer.okta.com/signup/).
* A Node.js Express framework sample application. This guide works with the sample application in the Sample code section below.

**Sample code**

* [Okta Token Inline Hook Example](https://glitch.com/edit/#!/okta-inlinehook-tokenhook)
* [Express Sample Applications for Okta](https://github.com/okta/samples-nodejs-express-4)

---

> The Token Inline Hook can be used to customize the authorization code flow that occurs between an application and the Okta org used for authentication.

Reviewer: ```suggestion
The Token Inline Hook can be used to customize the Authorization Code flow that occurs between an application and the Okta org used for authentication.
```

4. Add your external service URL, including the endpoint. For example, use your Glitch project name with the endpoint: `https://your-glitch-projectname.glitch.me/tokenHook`.

5. Include **Authentication field** and **Authentication secret** values. In this example:

    * **Authentication field** = `authorization`
    * **Authentication secret** = `Basic YWRtaW46c3VwZXJzZWNyZXQ=`

6. Click **Save**.

The Token Inline Hook is now set up with a status of active.

### Enable the Token Inline Hook

1. Navigate to **Security** > **API** > **Authorization Servers**.

2. Select a Custom Authorization Server from the list (usually **default**).

3. Select the **Access Policies** tab. Navigate to the Rule table and click the Edit icon next to the policy rule that will use the Inline hook. In most cases, edit the **Default Policy Rule** of the **Default Policy**.

4. From the **Use this inline hook** drop-down menu, select the Token Inline Hook you activated ("Patient Token Hook").

5. Click **Update Rule**.

The Token Inline Hook is now ready for triggering when the default policy rule is invoked from an authentication request.

## Extend the sample Express application

> The following code extends the local sample Node.js Express application to display the results of the Token Inline Hook claim addition. This step is optional. The Token Inline Hook is functional and the results of the implementation are available in the external service logs, as well as the System Log on your Okta org. But this extension is nice to have!

Reviewer: ```suggestion
The following code extends the local sample Node.js Express application to display the results of the Token Inline Hook claim addition. This step is optional. The Token Inline Hook is functional and the results of the implementation are shown in the external service logs, as well as in the system logs on your Okta org. But this extension is nice to have!
```

### Update the web server page

1. Navigate to your project folder `samples-nodejs-express-4` and continue to the `common` folder (`samples-nodejs-express-4/common`).
2. In an editor, open the `sample-web-server.js` page.
3. Locate the routing function `app.get('/profile'` and modify the function as in the code below. The Inline Token Hook code extension appears after the `const` declarations and before the `res.render` function.

This extension renders the [ID token](/docs/reference/api/oidc/#id-token), and if it contains the claim added by the Token Inline Hook, adds this claim to the attributes array. This array displays claims on the user's My Profile page.

<StackSelector snippet="extend-application" noSelector/>

## Test the Token Inline Hook

The Token Inline Hook is now ready for testing. You now have the following applications configured:

- The Okta-Hosted-Login sample application (`samples-nodejs-express-4`) is ready to authenticate users from your Okta org.
- The external service (Glitch.com project) is ready with code to receive and respond to an Okta Token Inline Hook call.

> - The Okta org is setup to call the external service when a Token Inline Hook is triggered by a user sign-in from the Okta-Hosted-Login sample application, and ready to receive a response.

Reviewer: ```suggestion
- The Okta org is set up to call the external service when a Token Inline Hook is triggered by a user sign-in from the Okta-Hosted-Login sample application, and ready to receive a response.
```

> >**Note:** Make sure you have users assigned to your application and at least one user is part of the [Patients data store](/docs/guides/token-inline-hook/#check-against-data-store/) in your Glitch application.

Reviewer: ```suggestion
>**Note:** Make sure you have users assigned to your application and at least one user is part of the [Patients data store](/docs/guides/token-inline-hook/#check-against-the-data-store/) in your Glitch application.
```

### Test your hook

1. Navigate to your sample application project folder (`samples-nodejs-express-4`).

2. Start your Okta-Hosted-Login server (`npm run okta-hosted-login-server`).

3. Navigate to your sample application (`http://localhost:8080`).

4. Navigate to your Glitch.com project and open the Console Logs window (**Tools** > **Logs**).

5. Return to your sample application and sign in with an Okta user who is NOT in the Patients data store.

    * The user should sign in as normal; only the user name should appear in the Glitch log window.
    * If you extended the sample application, click `My Profile` in the left-hand navigation pane. Only the the user info claims are included in the table.

6. Sign out of the sample application, and now sign in with an Okta user who IS in the Patients data store.

>     * The user should sign in as normal; however, this user should have a patient ID displayed in the Glitch console output, as well as a successful implementation record of the Token Inline Hook, available for review in your Okta org's System Log (**Reports** > **System Log**).

Reviewer: ```suggestion
    * The user should sign in as normal; however, this user should have a patient ID displayed in the Glitch console output, as well as a successful implementation record of the Token Inline Hook, available for review in your Okta org System Log (**Reports** > **System Log**).
```

    * If you extended the sample application, click `My Profile` in the left-hand navigation pane. The patient ID is added as part of the Claims table.

> **Note:** Review the [Token Inline Hooks troubleshooting](/docs/reference/token-hook/#troubleshooting) content or the [Troubleshooting hook implementations](/docs/guides/common-hook-set-up-steps/nodejs/main/#troubleshoot-hook-implementations) section for information on any difficulties.

## Next steps

Review the following guides to implement other Inline or Event Hook examples:

* [Event Hook](/docs/guides/event-hook-implementation/)
* [Registration Inline Hook](/docs/guides/registration-inline-hook/)
* [Password Import Inline Hook](/docs/guides/password-import-inline-hook/)

## See also

> 
> For further reference data on the Token Inline Hook, see: [Token Inline Hook](/docs/reference/token-hook/).

Reviewer: ```suggestion

For further reference data on the Token Inline Hook, see [Token Inline Hook](/docs/reference/token-hook/).
```
